File: 10_smart_pointers/python/bintree.py
class Node:
    printed = set()

    def __init__(self, value):
        self.parent = None
        self.left = None
        self.right = None
        self.value = value

    def __repr__(self):
        try:
            if id(self) in self.printed:
                return f"-"
            self.printed.add(id(self))
            spaces = " " * len(self.printed)
            s = f"Node\n{spaces}{spaces}parent: {self.parent}, left: {self.left}, right: {self.right}, v: {self.value}\n"
            return s
        except RecursionError:
            return "-"


if __name__ == '__main__':
    root = Node(0)
    node = Node(1)
    node.parent = root
    root.left = node
    node2 = Node(2)
    root.parent = node2

    print("Root:", root)
    Node.printed = set()
    print("Parent left:", root.left.parent.value)
    Node.printed = set()
    print("ids: {} {}".format(id(node2), id(root.parent)))
    del node2
    #root.parent = None
    print("Parent:   ", root.parent)

File: 10_smart_pointers/rust/bintree.rs
/// for illustration purposes only, do not use IRL!
use std::cell::RefCell; // a "runtime borrow-checker"
use std::fmt::Debug;
use std::rc::{Rc, Weak};

#[derive(Debug)]
struct Node<T> {
    parent: Option<Weak<RefCell<Node<T>>>>,
    left: Option<Rc<RefCell<Node<T>>>>,
    right: Option<Rc<RefCell<Node<T>>>>,
    value: T,
}

impl<T: Debug> Node<T> {
    fn new(value: T) -> Node<T> {
        Node {
            parent: None,
            left: None,
            right: None,
            value,
        }
    }

    fn print_parent(&self) {
        if let Some(parent) = &self.parent {
            if let Some(parent) = parent.upgrade() {
                println!("Parent: {:?}", parent.borrow().value);
            } else {
                println!("Parent: no more parent");
            }
        } else {
            println!("Parent: None");
        }
    }
}

fn main() {
    let root = Rc::new(RefCell::new(Node::new(0)));
    let node = Rc::new(RefCell::new(Node::new(1)));
    node.borrow_mut().parent = Some(Rc::downgrade(&root));

    let node2 = Rc::new(RefCell::new(Node::new(2)));

    root.borrow_mut().parent = Some(Rc::downgrade(&node2));
    root.borrow_mut().left = Some(node);

    println!("Root:   {:?}", *root.borrow());
    if let Some(left) = &root.borrow().left {
        &left.borrow().print_parent();
    };
    &root.borrow().print_parent();
    drop(node2);
    &root.borrow().print_parent();
}
